src/ast.rs

```diff
@@ -2834,91 +2298,44 @@
-            Category::branch => {
+            NodeRefMut::Branch(_branch) => {
                 // This field is a direct embedding of an AST value.
                 node.accept_mut(self, false);
-                return VisitResult::Continue(());
-            }
-            Category::list => {
-                // This field is an embedding of an array of (pointers to) ContentsNode.
-                // Parse as many as we can.
-                match node.typ() {
-                    Type::andor_job_list => self.populate_list::<AndorJobList>(
-                        node.as_mut_andor_job_list().unwrap(),
-                        false,
-                    ),
-                    Type::argument_list => self
-                        .populate_list::<ArgumentList>(node.as_mut_argument_list().unwrap(), false),
-                    Type::argument_or_redirection_list => self
-                        .populate_list::<ArgumentOrRedirectionList>(
-                            node.as_mut_argument_or_redirection_list().unwrap(),
-                            false,
-                        ),
-                    Type::case_item_list => self.populate_list::<CaseItemList>(
-                        node.as_mut_case_item_list().unwrap(),
-                        false,
-                    ),
-                    Type::elseif_clause_list => self.populate_list::<ElseifClauseList>(
-                        node.as_mut_elseif_clause_list().unwrap(),
-                        false,
-                    ),
-                    Type::job_conjunction_continuation_list => self
-                        .populate_list::<JobConjunctionContinuationList>(
-                            node.as_mut_job_conjunction_continuation_list().unwrap(),
-                            false,
-                        ),
-                    Type::job_continuation_list => self.populate_list::<JobContinuationList>(
-                        node.as_mut_job_continuation_list().unwrap(),
-                        false,
-                    ),
-                    Type::job_list => {
-                        self.populate_list::<JobList>(node.as_mut_job_list().unwrap(), false)
-                    }
-                    Type::variable_assignment_list => self.populate_list::<VariableAssignmentList>(
-                        node.as_mut_variable_assignment_list().unwrap(),
-                        false,
-                    ),
-                    _ => (),
-                }
             }
+            // This field is an embedding of an array of (pointers to) ContentsNode.
+            // Parse as many as we can.
+            NodeRefMut::List(list) => match list {
+                ListRefMut::AndorJobList(l) => self.populate_list(l, false),
+                ListRefMut::ArgumentList(l) => self.populate_list(l, false),
+                ListRefMut::ArgumentOrRedirectionList(l) => self.populate_list(l, false),
+                ListRefMut::CaseItemList(l) => self.populate_list(l, false),
```

Comment:
I wonder if we should define a generic type for lists, they're all the same.
We used to have that in C++ but I guess it goes against the exhaustive-enum-matching approach, so probably not.

Reply:
I already have plans that will make this entire function go away. Specifically:
```rs
trait Parse {
    fn parse(pop: &mut Populator<'_>) -> Self;
}
```
Leaves, branches, lists, and variants(?) will all have their own implementations of this trait, using logic relocated from `impl Populator`.
Branch impls will basically do this:
```rs
Self {
    parent: None,
    kw_else: Parse::parse(pop),
    semi_nl: Parse::parse(pop),
    body: Parse::parse(pop),
}
```
There will still be special handling for the "missing `end`" case that `VisitResult` handles, but it will look quite different.
Maybe towards the end of these changes, parsing could even use `Result` and the `?` operator. Then `Parse` impls for branches with a `KeywordEnd` would just need to attach context from their level of the tree before propagating.